### PR TITLE
DHIS2-3074 - Node leader election in cluster

### PR DESCRIPTION
**Pull request by ameenhere**
Uses redis to acquire a lock for leadership. The instance that has acquired lock is the leader. Jobconfiguration now has an additional column, a flag indicating whether the job has to be run in leader only or can be run any instance.

### CODE REVIEW

File: dhis-2/dhis-api/src/main/java/org/hisp/dhis/scheduling/SchedulingManager.java [hunk 1]
```diff
@@ -113,6 +113,13 @@
      * @param job The job to be executed
      */
     void executeJob( Runnable job );
+    
+    /**
+     * Schedule an ad hoc job without validation
+     * @param job The job to be scheduled
+     * @param startTime The time at which the job should start
+     */
+    void scheduleJob( Runnable job , Date startTime);
```
Comment (henninghaakonsen) [inline dhis-2/dhis-api/src/main/java/org/hisp/dhis/scheduling/SchedulingManager.java]: Is it possible to make this take in a JobConfiguration object? If it has inMemoryJob toggled there is no validation, but it runs through DefaultJobInstance which will make things more consistent.

It would also be nice if you rename it to scheduleJobAtTime(or something)

File: dhis-2/dhis-services/dhis-service-administration/src/main/java/org/hisp/dhis/leader/election/LeaderElectionJob.java [hunk 2]
```diff
@@ -0,0 +1,74 @@
...
+        try
+        {
+            leaderManager.electLeader();
+        }
+        catch ( Exception e )
+        {
+            notifier.notify( jobConfiguration, "Leader election failed:" + e.getMessage() );
+        }
+
+        notifier.notify( jobConfiguration, "Leader election completed" );
```
Comment (henninghaakonsen) [inline dhis-2/dhis-services/dhis-service-administration/src/main/java/org/hisp/dhis/leader/election/LeaderElectionJob.java]: You should propably use "public Notifier notify( JobConfiguration id, NotificationLevel level, String message, boolean completed )" instead so that if someone looks into the job in the notifier they will see that the job completed. In the exception, you can also set NotificationLevel to ERROR

File: dhis-2/dhis-services/dhis-service-core/src/main/java/org/hisp/dhis/scheduling/DefaultSchedulingManager.java [hunk 3]
```diff
@@ -78,21 +83,31 @@
...
 
     public void setTaskExecutor( AsyncListenableTaskExecutor jobExecutor )
     {
         this.jobExecutor = jobExecutor;
     }
 
+    @PostConstruct
+    public void init()
+    {
+        leaderManager.setSchedulingManager( this );
```
Comment (henninghaakonsen) [inline dhis-2/dhis-services/dhis-service-core/src/main/java/org/hisp/dhis/scheduling/DefaultSchedulingManager.java]: Doesn't it work autowiring SchedulingManager in the LeaderManager?

File: dhis-2/dhis-services/dhis-service-core/src/main/java/org/hisp/dhis/startup/SchedulerStart.java [hunk 4]
```diff
@@ -130,59 +157,98 @@ public void execute( )
...
             addAndScheduleJob( dataStatistics );
         }
 
-        if ( addDefaultJob( DEFAULT_VALIDATION_RESULTS_NOTIFICATION, jobConfigurations ) )
+        if ( verifyNoJobExist( DEFAULT_VALIDATION_RESULTS_NOTIFICATION, jobConfigurations ) )
         {
-            JobConfiguration validationResultNotification = new JobConfiguration( DEFAULT_VALIDATION_RESULTS_NOTIFICATION,
-                VALIDATION_RESULTS_NOTIFICATION, CRON_DAILY_7AM, null, false, true );
+            JobConfiguration validationResultNotification = new JobConfiguration(
+                DEFAULT_VALIDATION_RESULTS_NOTIFICATION, VALIDATION_RESULTS_NOTIFICATION, CRON_DAILY_7AM, null, false,
```
Comment (henninghaakonsen) [inline dhis-2/dhis-services/dhis-service-core/src/main/java/org/hisp/dhis/startup/SchedulerStart.java]: Codestyle

File: dhis-2/dhis-services/dhis-service-core/src/main/java/org/hisp/dhis/startup/SchedulerStart.java [hunk 5]
```diff
@@ -48,29 +48,44 @@
...
     private final String DEFAULT_DATA_SET_NOTIFICATION = "Dataset notification";
+
     private final String DEFAULT_REMOVE_EXPIRED_RESERVED_VALUES = "Remove expired reserved values";
 
+    private final String DEFAULT_LEADER_ELECTION = "Leader election in cluster";
+
     @Autowired
     private SystemSettingManager systemSettingManager;
 
+    private String redisEnabled;
```
Comment (henninghaakonsen) [inline dhis-2/dhis-services/dhis-service-core/src/main/java/org/hisp/dhis/startup/SchedulerStart.java]: Does these have to be methods and set by the bean?

File: dhis-2/dhis-services/dhis-service-core/src/main/java/org/hisp/dhis/startup/SchedulerStart.java [hunk 6]
```diff
@@ -48,29 +48,44 @@
...
  *
- * Reschedule old jobs and execute jobs which were scheduled when the server was not running.
+ * Reschedule old jobs and execute jobs which were scheduled when the server was
+ * not running.
  *
  * @author Henning Håkonsen
  */
-public class SchedulerStart
-    extends AbstractStartupRoutine
+public class SchedulerStart extends AbstractStartupRoutine
```
Comment (henninghaakonsen) [inline dhis-2/dhis-services/dhis-service-core/src/main/java/org/hisp/dhis/startup/SchedulerStart.java]: You should set the leaderJob here as well. Only doing it in the TableAlteror will not work if we create new Jobs here.

File: dhis-2/dhis-services/dhis-service-core/src/main/java/org/hisp/dhis/startup/SchedulerStart.java [hunk 7]
```diff
@@ -130,59 +157,98 @@ public void execute( )
...
+        }
+        
+    }
+
+    private void checkLeaderElectionJobConfiguration( List<JobConfiguration> jobConfigurations )
+    {
+        JobConfiguration leaderElectionJobConfiguration = jobConfigurations.stream()
+            .filter( jobConfiguration -> jobConfiguration.getName().equals( DEFAULT_LEADER_ELECTION ) ).findFirst()
+            .get();
+        if ( !leaderElectionJobConfiguration.getCronExpression()
```
Comment (henninghaakonsen) [inline dhis-2/dhis-services/dhis-service-core/src/main/java/org/hisp/dhis/startup/SchedulerStart.java]: Maybe just set the cron anyways? I guess it is equally fast to just set it. Unless it is configurable to some extent?

File: dhis-2/dhis-api/src/main/java/org/hisp/dhis/scheduling/JobConfiguration.java [hunk 8]
```diff
@@ -87,6 +87,8 @@
     private boolean inMemoryJob = false;
 
     private String userUid;
+    
+    private boolean leaderOnlyJob = false;
```
Comment (henninghaakonsen) [inline dhis-2/dhis-api/src/main/java/org/hisp/dhis/scheduling/JobConfiguration.java]: false here, but true in the .hbm file? What is the correct default value?

File: dhis-2/dhis-services/dhis-service-core/src/main/java/org/hisp/dhis/startup/TableAlteror.java [hunk 9]
```diff
@@ -1092,6 +1092,14 @@ public void execute()
         executeSql( "alter table userroledataset drop constraint fk_userroledataset_datasetid;" );
         executeSql( "alter table userroledataset drop constraint fk_userroledataset_userroleid;" );
         
+        // Update leaderOnlyJob flag in jobconfiguration
```
Comment (henninghaakonsen) [inline dhis-2/dhis-services/dhis-service-core/src/main/java/org/hisp/dhis/startup/TableAlteror.java]: Since the new attribute is "not-null" should you set all other rows to false, or given a certain JobType to true?

File: dhis-2/dhis-support/dhis-support-external/src/main/java/org/hisp/dhis/external/conf/ConfigurationKey.java [hunk 10]
```diff
@@ -83,6 +83,7 @@
     REDIS_PASSWORD( "redis.password", "", true ),
     REDIS_ENABLED( "redis.enabled", "false", false ),
     REDIS_USE_SSL( "redis.use.ssl", "false", false ),
+    LEADER_TIME_TO_LIVE( "leader.time.to.live.minutes", "2", false ),
```
Comment (henninghaakonsen) [inline dhis-2/dhis-support/dhis-support-external/src/main/java/org/hisp/dhis/external/conf/ConfigurationKey.java]: Is this related to the cron of the leader election job? If so, the "format"/cron of the job can be static

File: dhis-2/dhis-support/dhis-support-system/src/main/java/org/hisp/dhis/leader/election/NoOpLeaderManager.java [hunk 11]
```diff
@@ -0,0 +1,82 @@
...
+
+    private static final Log log = LogFactory.getLog( NoOpLeaderManager.class );
+
+    private String nodeId;
+
+    public NoOpLeaderManager()
+    {
+        this.nodeId = UUID.randomUUID().toString();
+        log.info( "Setting up noop leader manager on NodeId:" + this.nodeId );
+
```
Comment (henninghaakonsen) [inline dhis-2/dhis-support/dhis-support-system/src/main/java/org/hisp/dhis/leader/election/NoOpLeaderManager.java]: Remove line

File: dhis-2/dhis-api/src/main/java/org/hisp/dhis/leader/election/LeaderManager.java [hunk 12]
```diff
@@ -0,0 +1,51 @@
...
+
+/**
+ * Manages cluster leader node elections , renewals , revocations and to check
+ * whether the current instance is the leader in the cluster.
+ * 
+ * @author Ameen Mohamed
+ */
+public interface LeaderManager
+{
+    public void renewLeader();
```
Comment (henninghaakonsen) [inline dhis-2/dhis-api/src/main/java/org/hisp/dhis/leader/election/LeaderManager.java]: Remove "public"

File: dhis-2/dhis-api/src/main/java/org/hisp/dhis/leader/election/LeaderManager.java [hunk 13]
```diff
@@ -0,0 +1,51 @@
...
+
+    public void electLeader();
+
+    public void revokeLeader();
```
Comment (henninghaakonsen) [inline dhis-2/dhis-api/src/main/java/org/hisp/dhis/leader/election/LeaderManager.java]: This is not in use, maybe remove it or comment why it is there.

File: dhis-2/dhis-api/src/main/java/org/hisp/dhis/scheduling/SchedulingManager.java (hunk 1 above)
Comment (ameenhere) [reply dhis-2/dhis-api/src/main/java/org/hisp/dhis/scheduling/SchedulingManager.java]: It is more like an ad-hoc job. Similar to 
    /**
     * Execute an actual job without validation
     *
     * @param job The job to be executed
     */
    void executeJob( Runnable job );

I am not sure why we dont have a JobConfiguration parameter in the above method? I can add jobConfiguration to scheduleJob(). I also felt "scheduleJob(.., Date date) is kind of self explanatory. Maybe we can change it to scheduleJobWithStartTime()

File: dhis-2/dhis-services/dhis-service-administration/src/main/java/org/hisp/dhis/leader/election/LeaderElectionJob.java (hunk 2 above)
Comment (ameenhere) [reply dhis-2/dhis-services/dhis-service-administration/src/main/java/org/hisp/dhis/leader/election/LeaderElectionJob.java]: Got it!

File: dhis-2/dhis-services/dhis-service-core/src/main/java/org/hisp/dhis/scheduling/DefaultSchedulingManager.java (hunk 3 above)
Comment (ameenhere) [reply dhis-2/dhis-services/dhis-service-core/src/main/java/org/hisp/dhis/scheduling/DefaultSchedulingManager.java]: There is a circular dependency there. SchedulingManager already has LeaderManager autowired - to check if the current instance is a leader for a leaderOnlyJob. LeaderManager also needs SchedulingManager to schedule "Leader renewal adhoc jobs". There are a couple of ways to overcome this, one of them is what I implemented. Another way would be do add @Lazy initialization. Or maybe we would have to refactor it completely which I felt was unnecessary.

File: dhis-2/dhis-services/dhis-service-core/src/main/java/org/hisp/dhis/startup/SchedulerStart.java (hunk 4 above)
Comment (ameenhere) [reply dhis-2/dhis-services/dhis-service-core/src/main/java/org/hisp/dhis/startup/SchedulerStart.java]: The Dhis Code Formatter for eclipse formats it in this way. I will change it back to how it was then.

File: dhis-2/dhis-services/dhis-service-core/src/main/java/org/hisp/dhis/startup/TableAlteror.java (hunk 9 above)
Comment (ameenhere) [reply dhis-2/dhis-services/dhis-service-core/src/main/java/org/hisp/dhis/startup/TableAlteror.java]: Discussed with Stian, all the existing jobs (apart from the newly added leader election job), leaderOnlyJob will be true. Do you think that list is not exhaustive? Maybe we can change the scripts to set everything as true with 
```
update jobconfiguration set leaderonlyjob=true
```

and then explicitly set leaderelectionjob to false. Sounds ok?

File: dhis-2/dhis-support/dhis-support-external/src/main/java/org/hisp/dhis/external/conf/ConfigurationKey.java (hunk 10 above)
Comment (ameenhere) [reply dhis-2/dhis-support/dhis-support-external/src/main/java/org/hisp/dhis/external/conf/ConfigurationKey.java]: The format of the leader cron job is static in SchedulerStart.java
```
private final String LEADER_JOB_CRON_FORMAT = "0 0/%s * * * *";
```
We would want the minutes to be externally configured via the dhis.conf. The value given to "leader.time.to.live.minutes" will be replaced in the %s in the cron expression. Default would be every 2 minutes.

File: dhis-2/dhis-support/dhis-support-system/src/main/java/org/hisp/dhis/leader/election/NoOpLeaderManager.java (hunk 11 above)
Comment (ameenhere) [reply dhis-2/dhis-support/dhis-support-system/src/main/java/org/hisp/dhis/leader/election/NoOpLeaderManager.java]: Got it

File: dhis-2/dhis-api/src/main/java/org/hisp/dhis/leader/election/LeaderManager.java (hunk 12 above)
Comment (ameenhere) [reply dhis-2/dhis-api/src/main/java/org/hisp/dhis/leader/election/LeaderManager.java]: Got it

File: dhis-2/dhis-api/src/main/java/org/hisp/dhis/leader/election/LeaderManager.java (hunk 13 above)
Comment (ameenhere) [reply dhis-2/dhis-api/src/main/java/org/hisp/dhis/leader/election/LeaderManager.java]: Not in use currently. But highly likely to be used soon, when the clustering capabilities gets matured. I can add a comment indicating its not used currently.

File: dhis-2/dhis-services/dhis-service-core/src/main/java/org/hisp/dhis/startup/SchedulerStart.java (hunk 7 above)
Comment (ameenhere) [reply dhis-2/dhis-services/dhis-service-core/src/main/java/org/hisp/dhis/startup/SchedulerStart.java]: Got it.

File: dhis-2/dhis-services/dhis-service-core/src/main/java/org/hisp/dhis/startup/SchedulerStart.java (hunk 6 above)
Comment (ameenhere) [reply dhis-2/dhis-services/dhis-service-core/src/main/java/org/hisp/dhis/startup/SchedulerStart.java]: Got it.

File: dhis-2/dhis-api/src/main/java/org/hisp/dhis/scheduling/JobConfiguration.java (hunk 8 above)
Comment (ameenhere) [reply dhis-2/dhis-api/src/main/java/org/hisp/dhis/scheduling/JobConfiguration.java]: The general idea is to have leaderOnlyJob flag set to true for all jobs, which is why I set it to true in hbm. But after few rounds of testing, figured out that for the inMemoryJobs that are scheduled or created at runtime, there is always a new JobConfiguration object that is created. For such jobConfiguration it would make sense to default the flag to false, since the job has to be run by the instance in which the inMemoryJob is created irrespective of whether the instance is leader or not. Maybe I can change the default in hbm also to false. Since we have the TableAlteror setting all values to true anyway. Sounds ok?

File: dhis-2/dhis-api/src/main/java/org/hisp/dhis/leader/election/LeaderManager.java (hunk 13 above)
Comment (ameenhere) [reply dhis-2/dhis-api/src/main/java/org/hisp/dhis/leader/election/LeaderManager.java]: Removed it. Will add it when required. Based on the YAGNI principle. 👍

File: dhis-2/dhis-api/src/main/java/org/hisp/dhis/scheduling/JobConfiguration.java (hunk 8 above)
Comment (larshelge) [reply dhis-2/dhis-api/src/main/java/org/hisp/dhis/scheduling/JobConfiguration.java]: I think its a good convention and more natural to leave booleans with default = false. If leader only is the norm, could we reframe and call the property `parallelJob` ?

File: dhis-2/dhis-services/dhis-service-core/src/main/java/org/hisp/dhis/scheduling/DefaultSchedulingManager.java (hunk 3 above)
Comment (henninghaakonsen) [reply dhis-2/dhis-services/dhis-service-core/src/main/java/org/hisp/dhis/scheduling/DefaultSchedulingManager.java]: Ok - good.

File: dhis-2/dhis-services/dhis-service-core/src/main/java/org/hisp/dhis/startup/SchedulerStart.java (hunk 5 above)
Comment (henninghaakonsen) [reply dhis-2/dhis-services/dhis-service-core/src/main/java/org/hisp/dhis/startup/SchedulerStart.java]: I mean, is it necessary to set the fields? Could they be "static" like: 
private static String redisEnabled="redisEnabled"

File: dhis-2/dhis-services/dhis-service-core/src/main/java/org/hisp/dhis/startup/TableAlteror.java (hunk 9 above)
Comment (henninghaakonsen) [reply dhis-2/dhis-services/dhis-service-core/src/main/java/org/hisp/dhis/startup/TableAlteror.java]: Ok, I thought maybe some of the jobTypes were configurable. In that case users may have added other jobs which also would have to be set by this script.

File: dhis-2/dhis-support/dhis-support-external/src/main/java/org/hisp/dhis/external/conf/ConfigurationKey.java (hunk 10 above)
Comment (henninghaakonsen) [reply dhis-2/dhis-support/dhis-support-external/src/main/java/org/hisp/dhis/external/conf/ConfigurationKey.java]: Ok
